### PR TITLE
[NET-5017] APIGW Status Conditions for Gateway Policies

### PR DESCRIPTION
**Pull request by jm96441n**
Changes proposed in this PR:
- Add status conditions for gateway policies

How I've tested this PR:
1. clone the following repo https://github.com/jm96441n/consul-experiments/tree/main/k8s/jwts
1. from the `k8s/jwts` directory run the `start` script, this sets up a k8s cluster with an apigw, 3 services, a httproute, and a jwt provider named "local"
1. run `kubectl apply ./consul/gatewaypolicy-okta.yaml`
1. after that finishes run `kubectl describe gatewaypolicy gw-policy-2` to see all the statuses as invalid because the referenced provider does not exist
1. run `kubectl apply ./consul/okta.yaml` to create the `okta` jwt provider
1. run `kubectl describe gatewaypolicy gw-policy-2`  again to see the policy back in a healthy status

Checklist:
- [x] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: control-plane/api/v1alpha1/gatewaypolicy_types.go [hunk 1]
```diff
@@ -115,3 +115,21 @@ type GatewayJWTClaimVerification struct {
 	//   that this value matches.
 	Value string `json:"value"`
 }
+
+// GatewayPolicyStatus defines the observed state of the gateway
+type GatewayPolicyStatus struct {
```
Comment (jm96441n) [inline control-plane/api/v1alpha1/gatewaypolicy_types.go:120]: we needed a specific status type in order to add the status, similar to how the `Gateway` type in the k8s spec has a specific type for `GatewayStatus`

File: control-plane/api-gateway/binding/binder.go [hunk 2]
```diff
@@ -238,6 +243,19 @@ func (b *Binder) Snapshot() *Snapshot {
...
 			snapshot.Kubernetes.StatusUpdates.Add(&b.config.Gateway)
 		}
+
+		for idx, policy := range b.config.Policies {
+			policy := policy
+
+			var policyStatus v1alpha1.GatewayPolicyStatus
+
+			policyStatus.Conditions = policyValidation.Conditions(policy.Generation, idx)
+			// only mark the policy as needing a status update if there's a diff with it's old status
```
Comment (t-eckert) [inline control-plane/api-gateway/binding/binder.go:253]: "its" is possessive, "it's" is "it is".

```suggestion
			// only mark the policy as needing a status update if there's a diff with its old status
```

File: control-plane/api-gateway/binding/result.go [hunk 3]
```diff
@@ -582,3 +582,94 @@ func (l gatewayValidationResult) Conditions(generation int64, listenersInvalid b
...
+	return result.Conditions(generation)
+}
+
+func (g gatewayPolicyValidationResult) Conditions(generation int64) []metav1.Condition {
+	conditions := make([]metav1.Condition, 0)
+
+	conditions = append(conditions, g.acceptedCondition(generation))
+	conditions = append(conditions, g.resolvedRefsConditions(generation)...)
+	return conditions
+}
```
Comment (t-eckert) [inline control-plane/api-gateway/binding/result.go:611]: 🤷🏻 

```suggestion
func (g gatewayPolicyValidationResult) Conditions(generation int64) []metav1.Condition {
	return append([]metav1.Condition{g.acceptedCondition(generation)}, g.resolvedRefsConditions(generation)...)
}
```

**Review by t-eckert — APPROVED** [review]

File: charts/consul/templates/crd-gatewaypolicies.yaml [hunk 4]
```diff
@@ -199,42 +199,93 @@ spec:
...
+                - lastTransitionTime: "1970-01-01T00:00:00Z"
+                  message: Waiting for controller
+                  reason: Pending
+                  status: Unknown
+                  type: Accepted
+                - lastTransitionTime: "1970-01-01T00:00:00Z"
+                  message: Waiting for controller
+                  reason: Pending
+                  status: Unknown
+                  type: Programmed
```
Comment (andrewstucki) [inline charts/consul/templates/crd-gatewaypolicies.yaml:215]: So, this default doesn't match the valid conditions that the controller sets, I think you'll want to change it to `ResolvedRefs`

Comment (jm96441n) [reply charts/consul/templates/crd-gatewaypolicies.yaml:215]: good  catch! I had copied/pasted this and forgot to change it :facepalm:

File: control-plane/api-gateway/binding/binder.go (hunk 2 above)
Comment (jm96441n) [reply control-plane/api-gateway/binding/binder.go:253]: words are hard

**Pull request merged**